extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentPuller.java

```diff
@@ -86,11 +88,19 @@ public String getName()
       public InputStream openInputStream() throws IOException
       {
         try {
-          streamAcquired = true;
-          return s3Obj.getDataInputStream();
+          synchronized (inputStreamOpener) {
```

Comment:
@drcrallen I guess this ```synchronized``` block is needed because multiple threads in ```CacheScheduler``` might call this method at the same time.

However, ```storageObject.getDataInputStream()``` returns the same inputStream (https://bitbucket.org/jmurty/jets3t/src/default/src/org/jets3t/service/model/StorageObject.java#StorageObject.java-222), I also guess the above situation will cause an error.

So, what is the best solution for this? Maybe it's needed to download the full object content in some local filesystem, and returns different inputStreams per ```openInputStream()``` call? Or, is it better to use a single thread scheduledExecutor for CacheScheduler?

Reply:
I did this more to guarantee that the setting of `storageObject` and `streamAcquired` would be visible across threads. So more as a memory barrier than a concurrent access barrier.

Reply:
Not to say this is a "good idea" but that was the original intent. So if the object got passed around or referenced across threads, things wouldn't have weird race conditions in things like close. There should still only be one thread accessing the input stream.

Reply:
looks like there isn't such a thing here though >.< maybe the synchronization should be taken out.

Reply:
Thanks! I also tested that the returned input stream is not thread-safe. I removed the synchronized block and added a javadoc to this method.